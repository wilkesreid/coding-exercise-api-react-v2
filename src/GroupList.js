import React, { Component } from 'react'
import { Table, Icon } from 'semantic-ui-react'
import _sortBy from 'lodash/sortBy';
import _reverse from 'lodash/reverse';
import CsvImporter from './CsvImporter';
import ColumnIcon from './ColumnIcon';

const sortingMethods = {
    group_name: item => item.group_name.toLowerCase()
};

class GroupList extends Component {
    constructor(props) {
        super(props);
        this.state = {
            data: [],
            sortColumn: null,
            sortDirection: null,
        };
    }

    async componentDidMount() {
        let resp = await fetch("http://localhost:8000/api/groups");
        let data = await resp.json();
        this.setState({
            data: data.data,
        });
    }

    sortedItems() {
        if (this.state.sortColumn === null) {
            return this.state.data;
        }

        if (this.state.sortDirection === 'asc') {
            return _sortBy(this.state.data, sortingMethods[this.state.sortColumn]);
        }

        return _reverse(_sortBy(this.state.data, sortingMethods[this.state.sortColumn]));
    }

    isSortingBy(column) {
        return this.state.sortColumn === column;
    }

    sortBy(column) {
        if (this.state.sortColumn === column) {
            if (this.state.sortDirection === 'asc') {
                this.setState({
                    sortDirection: 'desc'
                });
            } else {
                this.setState({
                    sortColumn: null,
                    sortDirection: null
                });
            }
        } else {
            this.setState({
                sortColumn: column,
                sortDirection: 'asc'
            });
        }
    }

    render() {
        let items = this.sortedItems();

        return (
            <div>
            <Table>
                <Table.Header>
                    <Table.Row>
                        <Table.HeaderCell onClick={() => this.sortBy('group_name')} style={{cursor: 'pointer'}}>
                            Group Name
                            {this.isSortingBy('group_name') && <ColumnIcon direction={this.state.sortDirection} />}
                        </Table.HeaderCell>
                    </Table.Row>
                </Table.Header>

                <Table.Body>
                    {
                        items.map(({id, group_name}) => (
                            <Table.Row key={id}>
                                <Table.Cell>{ group_name }</Table.Cell>
                            </Table.Row>
                        ))
                    }
                </Table.Body>
            </Table>
            <CsvImporter
                model="groups"
                finished={() => {
                    fetch("http://localhost:8000/api/groups")
                    .then(resp => resp.json())
                    .then(data => {
                        console.log(data);
                        this.setState({
                            data: data.data,
                        });
                    });
                }}
                error={(err) => {
                    console.log('error', err);
                }}
            />
            </div>
        );
    }
}

export default GroupList;
