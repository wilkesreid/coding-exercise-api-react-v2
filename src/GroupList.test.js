import React from 'react';
import { shallow } from 'enzyme';
import GroupList from './GroupList';

let wrapper, data;

describe('<GroupList />', () => {

    beforeAll(() => {
        wrapper = shallow(<GroupList />)
        data = [{
            "id": 132,
            "group_name": "Developers",
            "updated_at": "2019-07-20 22:05:47",
            "created_at": "2019-07-20 22:05:47"
        }]

        wrapper.setState({ 'data' : data });
    });

    test('should match the snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });
})
