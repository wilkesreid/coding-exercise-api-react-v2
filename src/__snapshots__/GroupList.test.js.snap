// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GroupList /> should match the snapshot 1`] = `
<div>
  <Table
    as="table"
  >
    <TableHeader
      as="thead"
    >
      <TableRow
        as="tr"
        cellAs="td"
      >
        <TableHeaderCell
          as="th"
          onClick={[Function]}
          style={
            Object {
              "cursor": "pointer",
            }
          }
        >
          Group Name
        </TableHeaderCell>
      </TableRow>
    </TableHeader>
    <TableBody
      as="tbody"
    >
      <TableRow
        as="tr"
        cellAs="td"
        key="132"
      >
        <TableCell
          as="td"
        >
          Developers
        </TableCell>
      </TableRow>
    </TableBody>
  </Table>
  <CsvImporter
    error={[Function]}
    finished={[Function]}
    model="groups"
  />
</div>
`;

exports[`<ResultsList /> should match the snapshot 1`] = `
<div>
  <Table
    as="table"
    celled={true}
    padded={true}
  >
    <TableHeader
      as="thead"
    >
      <TableRow
        as="tr"
        cellAs="td"
      >
        <TableHeaderCell
          as="th"
          onClick={[Function]}
          singleLine={true}
          style={
            Object {
              "cursor": "pointer",
            }
          }
        >
          First Name
        </TableHeaderCell>
        <TableHeaderCell
          as="th"
          onClick={[Function]}
        >
          Last Name
        </TableHeaderCell>
        <TableHeaderCell
          as="th"
          onClick={[Function]}
        >
          Email
        </TableHeaderCell>
        <TableHeaderCell
          as="th"
          onClick={[Function]}
        >
          Status
        </TableHeaderCell>
      </TableRow>
    </TableHeader>
    <TableBody
      as="tbody"
    >
      <TableRow
        as="tr"
        cellAs="td"
        key="0"
      >
        <TableCell
          as="td"
          singleLine={true}
        >
          Macie
        </TableCell>
        <TableCell
          as="td"
          singleLine={true}
        >
          Emmerich
        </TableCell>
        <TableCell
          as="td"
          singleLine={true}
        >
          cremin.marjory@hotmail.com
        </TableCell>
        <TableCell
          as="td"
          singleLine={true}
        >
          active
        </TableCell>
      </TableRow>
    </TableBody>
  </Table>
  <CsvImporter
    error={[Function]}
    finished={[Function]}
    model="people"
  />
</div>
`;
